### PR TITLE
Add str method to datamodule

### LINKED ISSUE #9947: Support `str(datamodule)`
## 🚀 Feature

Add support

```python
print(str(MyDataModule()))
```

### Motivation

It currently prints:

```python
<__main__.MyDataModule object at 0x10284c970>
```

### Pitch

It could print the DataLoader structure:

```python
MyDataModule(
    train_dataloader: {"a": DataLoaderClass(batch_size=8, num_batches=16, num_workers=2), "b":  DataLoaderClass(batch_size=2, num_batches=16, num_workers=2)]
    val_dataloader: [DataLoaderClass(batch_size=3, num_batches=14, num_workers=0), DataLoaderClass(batch_size=8, num_batches=4, num_workers=0)]
    test_dataloader: DataLoaderClass(batch_size=4, num_batches=7, num_workers=2)
)
```

Or the number of batches per dataloader, similar to what was done in https://github.com/PyTorchLightning/pytorch-lightning/issues/5965

### Alternatives

Open to other ideas

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by MrWhatZitToYaa**
## What does this PR do?

Added a __str__ function to the DataModule in order to be able to print Datasets included in the module instead of an adress.

Fixes #9947 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20301.org.readthedocs.build/en/20301/

### CODE REVIEW

**Comment by MrWhatZitToYaa** [comment]
Will add documentation once code is fixed

**Comment by MrWhatZitToYaa** [comment]
I have been waiting for quite some time. How do I go on? Do I just wait some more until someone looks at this?

**Comment by lantiga** [comment]
hey @MrWhatZitToYaa thanks for taking this up and sorry for the wait

can you add a bit more detail on what do we think it's good printing from a datamodule?

the original design in #9947 was a bit different, as was #9967, but happy to consider alternatives after there's some discussion on the rationale

File: tests/tests_pytorch/core/test_datamodules.py [hunk 1]
```diff
@@ -510,3 +515,59 @@ def prepare_data(self):
...
+def test_datamodule_string_no_datasets():
+    dm = BoringDataModule()
+    del dm.random_full
+    expected_output = "No datasets are set up."
+    assert str(dm) == expected_output
+
+
+def test_datamodule_string_no_length():
+    dm = BoringDataModuleNoLen()
+    expected_output = "name=random_full, size=Unavailable\n"
```
[the remark this answers is not in the document]
Comment (MrWhatZitToYaa) [reply tests/tests_pytorch/core/test_datamodules.py]: I switched my code from the hardcoded \n to os.linesep

**Comment by MrWhatZitToYaa** [comment]
> hey @MrWhatZitToYaa thanks for taking this up and sorry for the wait
> 
> can you add a bit more detail on what do we think it's good printing from a datamodule?
> 
> the original design in #9947 was a bit different, as was #9967, but happy to consider alternatives after there's some discussion on the rationale

The idea originally came to me when I had to to use the Datamodule to implement some datasets we have in our lab. It was really annoying to deal with the fact that you never really know what you are dealing with and makes debugging really hard. I talked to some colleagues and they agreed that they ran into the same problem. That's why I would like to add a str functionality for the Datamodule: primarily for debugging / logging. 
I think the simplest approach is to scan the whole module for instances of datasets and then return a string where each line contains the name of the dataset and the length if it is available. This would be a big quality of life improvement when dealing with / constructing Datamodules.

File: src/lightning/pytorch/core/datamodule.py [hunk 2]
```diff
@@ -244,3 +245,34 @@ def load_from_checkpoint(
...
+                len(obj)
+                return True
+            except NotImplementedError:
+                return False
+
+        for attr_name in dir(self):
+            attr = getattr(self, attr_name)
+
+            # Get Dataset information
+            if isinstance(attr, Dataset):
```
Comment (lantiga) [inline src/lightning/pytorch/core/datamodule.py:269]: The problem with this is that there are no guarantees that any datasets actually exist as attributes.

I think a better design would be to leverage the contract, and get dataloaders through  `self.train_dataloader` (caveat: it may be a dataloader or an iterable of dataloaders), `self.val_dataloader`, etc, access their `dataset` property (https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) and produce a string from there in a similar way to the current way.

Also instead of ignoring non-Dataset subclasses in this case, we could call str from whatever is returned as dataset.

I think it's a valuable addition, wdyt?

**Review by lantiga — COMMENTED** [review]
Thanks for the contribution, I aligned it with the latest master and added a comment for a fix before we can merge. It should be quick.

File: src/lightning/pytorch/core/datamodule.py (hunk 2 above)
Comment (MrWhatZitToYaa) [reply src/lightning/pytorch/core/datamodule.py:262]: That is a good idea. I also had a similar idea in the beginning but discarded it because I thought it would add too much overhead. But I did some research an DataLoader should not add too much overhead to the actual dataset so we can use it in the str method.
Your suggestion would make the method more save, while still maintaining the same functionality, since one has to call prepare and setup anyway before one is able to access the dataset information.
I will try to implement your suggestion

**Comment by MrWhatZitToYaa** [comment]
The string method will now check if any of the 4 dataloaders are available and try to print information on them. I hope this is how you imagined it to work

**Review by lantiga — COMMENTED** [review]
That's correct. We're almost there, I propose we update the strings from

```
Train dataset: available=yes, size=unknown
Validation dataset: available=yes, size=64
Test dataset: available=no, size=unknown
Prediction dataset: available=yes, size=64
```

to

```
Train dataloader: size=NA
Validation dataloader: size=64
Test dataloader: None
Predict dataloader: size=64
```

summary:

- dataset -> dataloader
- Prediction -> Predict
- remove "available"
- unknown -> NA
- "available=no" -> None

what do you think?

**Comment by MrWhatZitToYaa** [comment]
Alright, sounds good to me
I made the changes so I assume it complies with your specification now

File: src/lightning/pytorch/core/datamodule.py [hunk 3]
```diff
@@ -244,3 +245,75 @@ def load_from_checkpoint(
...
+            return apply_to_collection(loader, DataLoader, retrieve_dataset_info)
+
+        def extract_loader_info(methods: list[tuple[str, str]]) -> dict:
+            """Helper function to extract information for each dataloader method."""
+            info: dict[str, Union[dataset_info, Iterable[dataset_info]]] = {}
+            for loader_name, func_name in methods:
+                loader_callback = getattr(self, func_name, None)
+
+                try:
+                    loader = loader_callback()
```
Comment (lantiga) [inline src/lightning/pytorch/core/datamodule.py:282]:
```suggestion
                    loader = loader_method()  # type: ignore
```

File: src/lightning/pytorch/core/datamodule.py [hunk 4]
```diff
@@ -244,3 +245,75 @@ def load_from_checkpoint(
...
+            loader: Union[DataLoader, Iterable[DataLoader]],
+        ) -> Union[dataset_info, Iterable[dataset_info]]:
+            """Helper function to compute dataset information."""
+            return apply_to_collection(loader, DataLoader, retrieve_dataset_info)
+
+        def extract_loader_info(methods: list[tuple[str, str]]) -> dict:
+            """Helper function to extract information for each dataloader method."""
+            info: dict[str, Union[dataset_info, Iterable[dataset_info]]] = {}
+            for loader_name, func_name in methods:
+                loader_callback = getattr(self, func_name, None)
```
Comment (lantiga) [inline src/lightning/pytorch/core/datamodule.py:279]:
```suggestion
                loader_method = getattr(self, func_name, None)
```

**Review by lantiga — APPROVED** [review]
